### PR TITLE
[ISSUE-1215]: Disk Eject E2E: RKE2 - Auto drive replacement - 1 drive failure out of 2 drives 

### LINKED ISSUE #1215: Disk Eject, test automation for the auto drive replacement - 1 drive failure out of 2 drives - SUCCESS
Automate whole scenario

### PR DESCRIPTION
**Pull request by mdutka-dell**
### Resolves #1215 

Automation of the scenario RKE2 - Auto drive replacement - 1 drive failure out of 2 drives - SUCCESS

## PR checklist
- [x] Add link to the issue
- [ ] Choose Project
- [ ] Choose PR label
- [ ] New unit tests added
- [x] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
image
image

### CODE REVIEW

**Review by katarzynakulpa — APPROVED** [review]

File: tests/e2e-test-framework/framework/utils.py [hunk 1]
```diff
@@ -803,16 +804,16 @@ def check_drive_cr_not_exist(self, drive_name: str, timeout: int = 120) -> bool:
...
-                    return True
+                    if not cr_existence:
+                        return True
                 else:
                     raise
             time.sleep(2)
-        logging.warning(
-            f"Drive CR '{drive_name}' still exists after {timeout} seconds timeout."
-        )
         return False
```
Comment (korzepadawid) [inline tests/e2e-test-framework/framework/utils.py:819]: question: did you run tox?

File: tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        failed_drive = drives[0]
+        health_drive = drives[1]
+        failed_volume = volumes[0]
+    # 2. simulate drive failure. Annotate drive used by pod with health=BAD
+        failed_drive_name = failed_drive["metadata"]["name"]
+        self.utils.annotate_custom_resource(
+            resource_name=failed_drive_name, 
+            resource_type="drives", 
+            annotation_key="health", 
+            annotation_value="BAD"
```
Comment (korzepadawid) [inline tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py:66]: nit: use consts

File: tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        assert self.utils.event_in(
+            resource_name=failed_drive_name,
+            reason=const.DRIVE_HEALTH_FAILURE,
+        ), f"event {const.DRIVE_HEALTH_FAILURE} for drive {failed_drive_name} not found"
+    # 6. annotate volume with release=done
+        self.utils.annotate_custom_resource(
+            resource_name=failed_volume_name, 
+            resource_type="volumes", 
+            annotation_key="release", 
+            annotation_value="done",
```
Comment (korzepadawid) [inline tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py:98]: nit: use consts

File: tests/e2e-test-framework/framework/utils.py [hunk 4]
```diff
@@ -775,23 +775,24 @@ def clear_pvc_and_pod(
...
                 assert self.wait_volume(
                     name=pvc.spec.volume_name,
-                    expected_usage=const.USAGE_RELEASED,
-                ), f"Volume: {pvc.spec.volume_name} failed to reach expected usage: {const.USAGE_RELEASED}"
-                logging.info(f"volume: {pvc.spec.volume_name} reach expected usage: {const.USAGE_RELEASED}")
+                    expected_usage=','.join([const.USAGE_RELEASED, const.USAGE_IN_USE]),
+                ), f"Volume: {pvc.spec.volume_name} failed to reach expected usage: {','.join([const.USAGE_RELEASED, const.USAGE_IN_USE])}"
+                logging.info(f"volume: {pvc.spec.volume_name} reach expected usage")
 
         time.sleep(30)
```
Comment (korzepadawid) [inline tests/e2e-test-framework/framework/utils.py:782]: question(non-blocking, non-related to the PR): Do we need to wait 30 seconds? It'd be better to wait for a specified state of a given resource.

File: tests/e2e-test-framework/framework/utils.py [hunk 5]
```diff
@@ -775,23 +775,24 @@ def clear_pvc_and_pod(
             for pvc in pvcs:
                 assert self.wait_volume(
                     name=pvc.spec.volume_name,
-                    expected_usage=const.USAGE_RELEASED,
-                ), f"Volume: {pvc.spec.volume_name} failed to reach expected usage: {const.USAGE_RELEASED}"
-                logging.info(f"volume: {pvc.spec.volume_name} reach expected usage: {const.USAGE_RELEASED}")
+                    expected_usage=','.join([const.USAGE_RELEASED, const.USAGE_IN_USE]),
+                ), f"Volume: {pvc.spec.volume_name} failed to reach expected usage: {','.join([const.USAGE_RELEASED, const.USAGE_IN_USE])}"
+                logging.info(f"volume: {pvc.spec.volume_name} reach expected usage")
```
Comment (korzepadawid) [inline tests/e2e-test-framework/framework/utils.py:780]: nit: let's reuse `expected_usage`

File: tests/e2e-test-framework/framework/utils.py (hunk 4 above)
Comment (korzepadawid) [inline tests/e2e-test-framework/framework/utils.py:782]: question(non-blocking, non-related to the PR): Do we need to wait 30 seconds? It'd be better to wait for a specified state of a given resource.

File: tests/e2e-test-framework/framework/utils.py (hunk 1 above)
Comment (korzepadawid) [inline tests/e2e-test-framework/framework/utils.py:819]: question: did you run tox?

File: tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py (hunk 2 above)
Comment (korzepadawid) [inline tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py:66]: nit: use consts

File: tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py (hunk 3 above)
Comment (korzepadawid) [inline tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py:98]: nit: use consts

File: tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        volumes = self.utils.list_volumes(pod_name=pod.metadata.name)
+        # get all drives
+        drives = []
+        for volume in volumes:
+            drive = self.utils.get_drive_cr(
+                volume_name=volume["metadata"]["name"], 
+                namespace=volume["metadata"]["namespace"])
+            drives.append(drive)
+        failed_drive = drives[0]
+        health_drive = drives[1]
```
Comment (korzepadawid) [inline tests/e2e-test-framework/tests/test_drive_replacement_multi_volumes_single_fail.py:58]: question: it should be called `healthy_drive`, shouldn't it?

**Review by korzepadawid — CHANGES_REQUESTED** [review]

File: tests/e2e-test-framework/framework/utils.py (hunk 4 above)
Comment (mdutka-dell) [reply tests/e2e-test-framework/framework/utils.py:782]: to be done if needed in next PRs

File: tests/e2e-test-framework/framework/utils.py (hunk 1 above)
Comment (mdutka-dell) [reply tests/e2e-test-framework/framework/utils.py]: done

**Review by korzepadawid — APPROVED** [review]
LGTM

**Pull request merged**